src/pysdmx/model/__base.py

```diff
@@ -0,0 +1,212 @@
+from datetime import datetime
+from typing import Any, Dict, Optional, Sequence, Union
+
+from msgspec import Struct
+
+
+class Annotation(Struct, frozen=True, omit_defaults=True):
+    """Annotation class.
+
+    It is used to convey extra information to describe any
+    SDMX construct.
+    This information may be in the form of a URL reference and/or
+    a multilingual text (represented by the association to
+    InternationalString).
+
+    Attributes:
+        id: The identifier of the annotation.
+        title: The title of the annotation.
+        text: The text of the annotation.
+        url: The URL of the annotation.
+        type: The type of the annotation.
+    """
+
+    id: Optional[str] = None
+    title: Optional[str] = None
+    text: Optional[str] = None
+    url: Optional[str] = None
+    type: Optional[str] = None
+
+    def __str__(self) -> str:
+        """Returns a human-friendly description."""
+        out = []
+        for k in self.__annotations__.keys():
+            v = self.__getattribute__(k)
+            if v:
+                out.append(f"{k}={str(v)}")
+        return ", ".join(out)
+
+
+class AnnotableArtefact(Struct, frozen=True, omit_defaults=True):
+    """Annotable Artefact class.
+
+    Superclass of all SDMX artefacts.
+    Contains the list of annotations.
+
+    Attributes:
+        annotations: The list of annotations attached to the artefact.
+    """
+
+    annotations: Sequence[Annotation] = ()
+
+    @classmethod
+    def __all_annotations(cls) -> Dict[str, Any]:
+        class_attributes = {}
+        for c in cls.__mro__:
+            if "__annotations__" in c.__dict__:
+                class_attributes.update(c.__annotations__)
+        return dict(reversed(list(class_attributes.items())))
+
+    def __str__(self) -> str:
+        """Returns a human-friendly description."""
+        out = []
+        for k in self.__all_annotations().keys():
+            v = self.__getattribute__(k)
+            if v:
+                out.append(f"{k}={str(v)}")
+        return ", ".join(out)
+
+
+class IdentifiableArtefact(AnnotableArtefact, frozen=True, omit_defaults=True):
+    """Identifiable Artefact class.
+
+    Provides identity to all derived classes. It also provides annotations
+    to derive classes because it is a subclass of AnnotableArtefact.
+
+    Attributes:
+        id: The identifier of the artefact.
+        uri: The URI of the artefact.
+        urn: The URN of the artefact.
+    """
+
+    id: str = ""
```

Comment:
Why is the id defaulting to an empty string?

Reply:
This is to prevent mypy errors:

![image](https://github.com/bis-med-it/pysdmx/assets/73304358/fd211372-e58e-44e7-8ea8-45eeb239a2fc)

We could add a misc ignore to prevent this if needed. Defaulting id to an empty string may raise some issues for the final user, so it´s better to enforce the use of it. Do you agree?

Reply:
Need to add also kw_only=True to AnnotableArtefact to prevent misuse of the arguments to init all classes, it is also inherited by the subclasses.

Reply:
> We could add a misc ignore to prevent this if needed. Defaulting id to an empty string may raise some issues for the final user, so it´s better to enforce the use of it. Do you agree? @sosna
Yes, indeed, I agree. We should add a post init check to make sure the id has not been left empty